ckanext/dalrrd_emc_dcpr/plugin.py

```diff
@@ -15,3 +21,83 @@ def update_config(self, config_):
 
     def get_commands(self):
         return [test_ckan_cmd]
+
+    # IAuthFunctions
+
+    def get_auth_functions(self) -> typing.Dict[str, typing.Callable]:
+        return {
+            "package_publish": authorize_package_publish,
+        }
+
+    def get_actions(self) -> typing.Dict[str, typing.Callable]:
+        return {
+            "package_create": package_create,
+            "package_update": package_update,
+            "package_patch": package_patch,
+        }
+
+
+def authorize_package_publish(
+    context: typing.Dict, data_dict: typing.Optional[typing.Dict] = None
+) -> typing.Dict[str, bool]:
+    # in here we can inspect the context and the data_dict in order to
+    # decide whether to authorize access or not
+    user_name = context.get("user")
+    owner_org = data_dict.get("owner_org") or data_dict.get("group_id")
+
+    members = toolkit.get_action("member_list")(
+        data_dict={"id": owner_org, "object_type": "user"}
+    )
+    admin_member_ids = [
+        member_tuple[0] for member_tuple in members if member_tuple[2] == "Admin"
+    ]
+    convert_user_name_or_id_to_id = toolkit.get_converter(
+        "convert_user_name_or_id_to_id"
+    )
+    user_id = convert_user_name_or_id_to_id(user_name, context)
+
+    if user_id in admin_member_ids:
+        return {"success": True}
+    else:
+        logger.debug(f"Inside authorize_package_publish function: {locals()}")
+        return {"success": False, "msg": "You are not authorized to publish a package"}
+
+
+# IActions
+
+
+@toolkit.chained_action
+def package_create(original_action, context, data_dict):
+    """
+    Intercepts the core `package_create` action to check if package
+     is being published after being created.
+    """
+    return package_publish_check(original_action, context, data_dict)
+
+
+@toolkit.chained_action
+def package_update(original_action, context, data_dict):
+    """
+    Intercepts the core `package_update` action to check if package is being published.
+    """
+    return package_publish_check(original_action, context, data_dict)
+
+
+@toolkit.chained_action
+def package_patch(original_action, context, data_dict):
+    """
+    Intercepts the core `package_patch` action to check if package is being published.
+    """
+    return package_publish_check(original_action, context, data_dict)
+
+
+def package_publish_check(action, context, data):
+    remains_private = toolkit.asbool(data.get("private", True))
+    result = None
+    if remains_private:
+        # make sure private state is updated, then assign result
+        data.update({"private": remains_private})
+        result = action(context, data)
+    else:
+        toolkit.check_access("package_publish", context, data)
```

Comment:
This is incomplete. So you check the permission, then what?

If the user is authorized to make the change, then this code must include a call to the `action`.

Reply:
You have a point, but why does everything works as expected!

Reply:
It seems to me that you probably are not testing this thoroughly.

What I mean is, if you are hiding the `visible` input field in the UI and then use the same UI to test this, then you will never reach the path where your user is trying to do something which cannot be done - this is actually the point of us modifying the UI in order to prevent the user from making mistakes.

How can you ever reach a state where your user is not authorized to change a package's visibility if you are using the UI, that hides that field from users which are not allowed to do it?

Test this by using the CKAN REST API - this will not hide anything from your user. Provide the `private` field to the API and I'm betting you will be able to find the error in this code.